feat: add version commit message management

chenghao-guo opened the pull request:
closes https://github.com/lancedb/lance/issues/4027

Related with https://github.com/lancedb/lance/issues/4181. We are going to break down the implementation into 2 phases. This work is the phase 1 for  #4181 . We added the properties in Transaction proto, and currently support high level API like write_dataset/add_columns to update the properties when performing the action.  The phase 2 PR will focus on Java API and will be proposed a few days later including the low level API like commit and expose all these API to Java binding methods which required by the fluss integration.


## Distinction from Tags
This feature is distinct from the existing tags functionality:
- Tags are stable pointers to milestone versions (like a git tag).
- Commit Messages are descriptive strings explaining the change for every version (like a git commit message).

## Updated Implementation Details: (July 15th, 2025)
The core of this feature is a new properties parameter, which accepts a dictionary of string key-value pairs during write operations. For convenience, a commit_message parameter is also provided as a shortcut for the common use case of adding a descriptive message. It simply sets the commit_message key within the properties dictionary in python level. Inside the rust, we always use the properties to pass through. commit_message is just a specific key of properties for this feature.

Metadata is attached at the time of a transaction via new optional parameters on write operations. It cannot be modified after a version has been committed.
- `lance.write_dataset(..., commit_message: Optional[str] = None, properties: Optional[Dict[str, str]] = None)`
- `LanceDataset.add_columns(..., commit_message: Optional[str] = None, properties: Optional[Dict[str, str]] = None)`
- (Other write operations like merge_insert follow a similar pattern and will be supported in the future)


New methods have been added to retrieve this information:
- `LanceDataset.list_commit_messages(max_transactions: int = 10) -> List[Tuple[int, Optional[str]]]`: Retrieves a list of recent versions and their associated commit messages.
- `LanceDataset.get_transaction_properties(version: int) -> Optional[Dict[str, str]]`: Retrieves the full dictionary of properties for a specific dataset version.

## Remaining work
1. Lower level api like commit(). I will need to discuss with Jack on where we can set the properties in this thread. After some guidance, I will propose the Phase2 PR.

For low level commit API like def commit(.....), shall we add a new parameter properties: ```Optional[Dict[str, str]]```  in the commit method, or we add the properties when we building the ``` operation: Union[LanceOperation.BaseOperation, Transaction]```?

2. expose all above methods in Java API for fluss integration. This should be a quick task after the PR merged.



## (Deprecated)Previous Implementation Details

~~1. `lance.write_dataset(..., commit_message: str)`~~
~~A new optional parameter `commit_message` has been added to the `write_dataset` function. This allows associating a message with the initial version of the dataset upon creation or with a new version during an `append` operation.~~

~~For other methods like add_columns/merge_insert etc, additional parameter will be added in the future when writing. At this time we could just use the commit_message method  below to amend a commit message after add_columns to achieve the same, examples have been added in `test_dataset.py`~~

~~2.  experimental version management api by overwriting existing manifest config~~
~~- `LanceDataset.commit_message(message: str, version: Optional[int] = None)`~~
~~This new method allows adding or updating a commit message for a specific version of the dataset. If `version` is not provided, the message is applied to the latest version~~
~~-  `LanceDataset.delete_message(version: Optional[int] = None)`~~
~~This method removes the commit message from a specific version. If `version` is not provided, it targets the latest version~~
~~-  `LanceDataset.list_commit_messages() -> List[Tuple[int, Optional[str]]]`~~
~~This method retrieves a list of all versions and their associated commit messages. It returns a list of tuples, where each tuple contains the version number and the commit message (or `None` if no message is set)~~

wjones127 commented on rust/lance/src/dataset.rs:
```diff
@@ -1848,6 +1930,75 @@ impl Default for ManifestWriteConfig {
+    }
+}
+
+/// modify the existing manifest file and commit to specific version,
+/// experimentally API designed for commit_message management,
+/// should not be called in concurrent situation
+pub(crate) async fn modify_existing_manifest_file(
```
manifest files should be considered immutable. We should not overwrite them.

jackye1995 replied:
I am just a bit worried with the one you are proposing above, "How about putting a message file under the versions/ folder, and then set up links to different versions. Maybe like a key-value structure.", if that would be too complicated. After all, I like the simplicity of just putting commit message in the config map which is a free-form map for arbitrary information. The only issue right now seems to be that this can only work for new commits, and also the message cannot be modified. But I think these restrictions are actually fine. For historical commits, as I said above, we just need some similar features for tags.

chenghao-guo replied:
> > , "The config field is meant for values that should be automatically carried forward to new versions, like configuration".
> 
> Yeah I totally agree. We are discussing below that we could store this information in the transaction file, what do you think of that?

Yes, I've studied the discussion below and completely agree with the direction. To be transparent, I'm still learning the intricacies of our transaction implementation. Before proposing any changes, I plan to carefully review the existing protobuf message structures and previous design decisions to understand the original intentions. Once I grasp the core principles, I'll explore how to adapt this solution into our transaction logic while respecting the established patterns. Would appreciate guidance if you spot any gaps in my approach!"

wjones127 commented on rust/lance/src/dataset.rs:
```diff
@@ -568,6 +570,85 @@ impl Dataset {
+            )
+            .await?;
+
+            let commit_message = manifest.config.get(COMMIT_MESSAGE_KEY).cloned();
```
The `config` field is meant for values that should be automatically carried forward to new versions, like configuration. I don't think the commit message is appropriate for this field.

wjones127 reviewed (changes requested):
I like the idea, but I think this is in the wrong place.

Manifests are meant to be immutable. (Breaking this immutability can [cause serious bugs](https://github.com/lancedb/lance/issues/1356).) This is why when we implemented tags, we put them in a separate file / structure that was designed to be mutated. I think you should consider doing the same for the commit messages.

BTW I think you should also think about retention of messages. We made tags prevent a version from being deleted. But would we want the same of commit messages? I would think we'd want some way to preserve at least the message and transaction, but not have to keep around the old data. I was thinking of a way to do this in this issue: https://github.com/lancedb/lance/issues/3563

jackye1995 commented on rust/lance/src/dataset/transaction.rs:
```diff
@@ -1381,6 +1385,10 @@ impl Transaction {
         };
 
+        if let Some(commit_msg) = &config.commit_message {
+            manifest.update_config(vec![(COMMIT_MESSAGE_KEY.to_string(), commit_msg.clone())]);
```
I am thinking if we could make this more generic, rather than setting a message, typically there is a bunch of information that a commit might want to carry (e.g. engine name that produced the commit, the engine ID, any associated pipeline ID, triggered by user name, etc.) 

I feel it makes more sense to actually allow any write operation to just pass in a key value map, and this is recorded as a part of the config. And we can have commonly used keys such as `commit_message`. What do we think? @wjones127 @chenghao-guo

wjones127 replied:
Arguably, this belongs in the transaction protobuf message.

jackye1995 replied:
> this belongs in the transaction protobuf message.

That makes sense to me

jackye1995 commented on rust/lance/src/dataset/transaction.rs:
```diff
@@ -1381,6 +1385,10 @@ impl Transaction {
+        if let Some(commit_msg) = &config.commit_message {
+            manifest.update_config(vec![(COMMIT_MESSAGE_KEY.to_string(), commit_msg.clone())]);
+        }
+
```
Another point to think about is that, this could be considered really as a composite transaction of a write + a ConfigUpdate. If we finish the transaction v2 work, that could potentially achieve the same result.

jackye1995 commented on rust/lance/src/dataset.rs:
```diff
@@ -1455,6 +1482,56 @@ impl Dataset {
+    /// # Returns
+    ///
+    /// Returns a vector of tuples containing version number and commit message (if available).
+    pub async fn list_commit_messages(
```
I think it will be more useful to just have methods like `list_transactions`  and then you could build `list_commit_message` on that.

jackye1995 replied:
> it seems not easy to extract that properties from python level, since it was already str and looks like a json.

I think the core problem is that it's returning the transaction as a string? (that's also probably why it's in debug module)

if we have a list_transactions function that returns proper `Transaction`, would that fix the issue?

And I think having `list_transactions` in either python-ffi or rust works, I was thinking `list_transactions` would be pretty handy so maybe worth having that in rust directly.

jackye1995 commented on rust/lance/src/dataset.rs:
```diff
@@ -1455,6 +1482,56 @@ impl Dataset {
+                results.push((version, None));
+            }
+
+            // If we've reached the maximum number of transactions, stop
```
these code feels very AI-written lol, can we remove the unnecessary inline comments?

chenghao-guo replied:
Sure

jackye1995 commented on rust/lance/src/dataset/write/insert.rs:
```diff
@@ -265,7 +265,7 @@ impl<'a> InsertBuilder<'a> {
         });
 
-        Ok(Transaction::new(
+        let mut transaction = Transaction::new(
```
since we added properties to the transaction proto, I think it makes more sense to just put properties as a part of the constructor.

jackye1995 replied:
In that case, seems like we should develop a `TransactionBuilder`? But overall I think it's better to either use a builder or update the `new`, instead of making the struct mutable for the sake of setting the optional field.

jackye1995 commented on rust/lance/src/dataset/write/commit.rs:
```diff
@@ -442,6 +452,7 @@ impl<'a> CommitBuilder<'a> {
             read_version,
             blobs_op,
             tag: None,
+            properties: None,
```
add a TODO for handing this

chenghao-guo replied:
Got it

chenghao-guo commented on protos/transaction.proto:
```diff
@@ -28,6 +28,9 @@ message Transaction {
   string tag = 3;
 
+  // Extra metadata for the transaction, such as commit message
+  map<string, string> properties = 4;
```
Sure, let's use __lance_commit_message as internal key.  I am going to define it as a const VALUE in transaction.rs . Maybe we use 
```pub const LANCE_COMMIT_MESSAGE_KEY: &str = "__lance_commit_message"``` ?

how do you think about that?

jackye1995 replied:
sounds good, just make sure it is also documented here.

chenghao-guo commented on rust/lance/src/dataset.rs:
```diff
@@ -665,6 +665,33 @@ impl Dataset {
+    ///
+    /// A HashMap containing the properties of the Transaction, or None if the transaction
+    /// doesn't exist or has no properties.
+    pub async fn get_transaction_properties(
```
Looks like Will already implemented the read_transaction() method in line:658 to 665.

And in the get_transaction_properties, I just reused the read_transaction method to retrieve its properties. So maybe it already follow that? Or shall I change the method name from "get_transaction_properties" to "read_transaction_properties", so that it looks better, and we can think of there should be a read_transaction method already?

jackye1995 replied:
> Looks like Will already implemented the read_transaction() method in line:658 to 665.

Yeah I was saying that because of this, it seems like what is missing is just you checkout to the specific version and read the transaction. We can wrap that in a `get_transaction`, or just call the 2 steps separately, I am fine with both. But overall my comment was that can we avoid having a specific function just for getting a particular field in the transaction?

chenghao-guo replied:
Got it. I agree with that, I think I can expose the below function
`pub async fn read_transaction(&self) -> Result<Option<Transaction>>`

to a python (later Java) interface, then it shall be fine. Then I can simply remove the get_transaction_properties

jackye1995 replied:
sounds good!

chenghao-guo commented:
Hi @jackye1995 , hope you're doing great! 

Following your valuable suggestions on the transaction API design, I've implemented the requested updates in the PR:
1. Introduced `TransactionBuilder` for optional property construction
2. Consolidated transaction access methods per our discussion like read_transaction and list_transactions which will return structured transaction. I did some changes on current Rust→Python Transaction conversion, it may incur partial information loss but likes enough for this case. Future plan can be supported if any other information required. Other changes have been documented in detail within the PR description. When your schedule permits, I'd appreciate your review of the implementation approach.  Java API will be added later after the PR accepted for fluss integration.

Please don't hesitate to reach out if you have any questions or require additional clarification on specific implementation choices.

wjones127 commented on rust/lance/src/dataset/transaction.rs:
```diff
@@ -1018,17 +1019,58 @@ impl Operation {
-impl Transaction {
-    pub fn new_from_version(read_version: u64, operation: Operation) -> Self {
-        let uuid = uuid::Uuid::new_v4().hyphenated().to_string();
+/// Add TransactionBuilder for flexibly setting option without using `mut`
+pub struct TransactionBuilder {
```
👍

wjones127 commented on rust/lance/src/dataset/schema_evolution.rs:
```diff
@@ -296,6 +297,7 @@ pub(super) async fn add_columns(
     transforms: NewColumnTransform,
     read_columns: Option<Vec<String>>,
     batch_size: Option<u32>,
+    properties: Option<HashMap<String, String>>,
```
I'm not sure I want to make you do this, but now that we have 3 optional arguments, we should really move this to a builder. This is another breaking change in the Rust API (adding another parameter) and it made you change all other other calls in the unit tests.

wjones127 replied:
If you instead created a `add_columns_builder()` method that returned a builder, you could add the new `properties` parameter to the builder and leave this method as-is.

chenghao-guo replied:
Sure, let me make some changes to adapt this.

chenghao-guo replied:
remove the add_columns commit_message support as this seems not perfect. Make the PR focus on transaction properties protbuf  first.

wjones127 reviewed (commented):
This PR has breaking changes for both the `debug::list_transactions` function and the `Dataset::add_columns()` method. If you must keep them, you need to label this PR as a breaking change and follow the steps in https://github.com/lancedb/lance/blob/main/release_process.md#breaking-change-policy.

However, I'd prefer if this PR didn't make any breaking changes.

chenghao-guo commented on python/python/lance/debug.py:
```diff
@@ -8,4 +8,4 @@
 from .lance import format_manifest as format_manifest
 from .lance import format_schema as format_schema
-from .lance import list_transactions as list_transactions
+from .lance import list_transactions_for_debug as list_transactions_for_debug
```
I will rename the new method to avoid breaking the previous debug::list_transactions

chenghao-guo commented:
Hi @wjones127 @jackye1995 , Greetings!  Could you please take an additional review to check if any further modification required at your convenience during your office hours? Thanks in advance.

jackye1995 commented on python/src/transaction.rs:
```diff
@@ -241,7 +242,128 @@ impl<'py> IntoPyObject<'py> for PyLance<&Operation> {
+            }
+            _ => {
+                //use BaseOperation as a fallback
+                let base_op = namespace.getattr("BaseOperation")?;
```
why use a base? I think we should just fail here

chenghao-guo replied:
Looks like previous code snippet was inconsistent with the Transaction Operation Enum. This will directly cause "not yet implemented failure"  if new Operation added in rust, but not added here.  While previous code doesn't have to convert the  whole Transaction object back to python, this seems to be a "soft limit" that that we just parse part of a Transaction conversion to Python. 

 In this PR, I have tried to supplement a lot of previous missed Operation, but I am afraid of that I may still miss some operation, that's the reason I replace the todo with a 'fallback' because python all inherit BasicOperation.   Even some code miss the implementation in transaction conversion, it will not cause directly failure.
 
 I can definitely change back to "todo!". This will constraint that every Operation change in the rust Transaction may also need implementation in python conversion.  The "soft limit" of a Transaction pyobject conversion becomes a "hard limit" for future contributors. Not sure on @wjones127 's opinion on this.

jackye1995 commented on rust/lance/src/dataset/write.rs:
```diff
@@ -201,6 +202,10 @@ pub struct WriteParams {
+
+    /// Configuration key-value pairs for this write operation.
+    /// This can include commit messages, engine information, etc.
+    pub properties: Option<HashMap<String, String>>,
```
I think this should be called `transaction_properties` to be specific, the comment should also describe that this properties map will be persisted as a part of the transaction object.

chenghao-guo replied:
Done. To avoid confusion,  I have refracted all changes to transaction_properties.

jackye1995 commented on rust/lance/src/dataset/write/commit.rs:
```diff
@@ -442,6 +452,8 @@ impl<'a> CommitBuilder<'a> {
             read_version,
             blobs_op,
             tag: None,
+            //TODO: handle batch transaction merges in the future
```
please also add a corresponding tracking issue.

jackye1995 reviewed (commented):
mostly looks good to me, just some nit comments

chenghao-guo commented on python/python/lance/dataset.py:
```diff
@@ -2742,31 +2759,47 @@ def config(self) -> dict[str, str]:
-    def sql(self, sql: str) -> "SqlQueryBuilder":
-        """Execute SQL query on the dataset.
+    def read_transaction(self, version: int) -> Optional[Transaction]:
+        """Get properties from a Transaction at a specific version.
```
Looks like due to git conflict resolve, and git diff looks like messed up.

I have just adjusted the format manually, should looks clean now.

jackye1995 reviewed (approved):
looks good to me. @wjones127 could you take a look again when have time?

chenghao-guo commented:
Rebased daily to resolve conflicts

chenghao-guo commented:
> looks good to me. @wjones127 could you take a look again when have time?

Hi @wjones127 , hope you are doing great! When you have a moment to spare, I'd be grateful if you could kindly take a look at this PR. Would you mind sharing your valuable insights on whether it needs any further refinement?  No rush at all - just whenever works best for your schedule.

chenghao-guo commented:
Emm, cargo clippy suddenly fails due to  `clippy::large_futures size limitation` after rebasing several commits. Probably reaching some limitation in clippy ? Will take a futher look next week.

chenghao-guo commented:
Hi @wjones127 , hope you're doing well! When you have a moment, I'd appreciate if you could review this PR as there was a requested change from you previously. It shall contain no more breaking changes now, and I'm hoping to get your approval to merge when convenient for your schedule.

chenghao-guo commented on python/python/lance/dataset.py:
```diff
@@ -2778,6 +2780,48 @@ def config(self) -> dict[str, str]:
+        """
+        return self._ds.read_transaction(version)
+
+    def get_transactions(self) -> Optional[List[Transaction]]:
```
sorry, missed, just  fixed

wjones127 commented on python/python/lance/dataset.py:
```diff
@@ -2778,6 +2780,48 @@ def config(self) -> dict[str, str]:
+        """
+        return self._ds.read_transaction(version)
+
+    def get_transactions(self, max_transactions=10) -> Optional[List[Transaction]]:
+        return self._ds.get_transactions(max_transactions)
```
I wonder if this should be called `recent_transaction` to be more descriptive?

wjones127 commented on python/python/lance/dataset.py:
```diff
@@ -2997,6 +3041,7 @@ class Transaction:
     operation: LanceOperation.BaseOperation
     uuid: str = dataclasses.field(default_factory=lambda: str(uuid.uuid4()))
     blobs_op: Optional[LanceOperation.BaseOperation] = None
+    transaction_properties: Optional[Dict[str, str]] = None
```
I think this can just default to an empty dict.
```suggestion
    transaction_properties: Dict[str, str] = dataclasses.field(default_factory=dict)
```

chenghao-guo replied:
Thanks for the review! Good call — I’ll update this to default to an empty dict.

wjones127 commented on python/python/tests/test_dataset.py:
```diff
@@ -3728,3 +3728,112 @@ def test_dataset_sql(tmp_path: Path):
+
+    # Test retrieving properties - should return None
+    transaction = dataset.read_transaction(1)
+    assert transaction.transaction_properties is None
```
```suggestion
    assert transaction.transaction_properties == {}
```

wjones127 commented on python/python/tests/test_dataset.py:
```diff
@@ -3728,3 +3728,112 @@ def test_dataset_sql(tmp_path: Path):
+    assert transaction.transaction_properties is None
+
+
+def test_get_transactions_properties(tmp_path):
```
Are you going to write a test where you construct a transaction in Python with a message and commit it? That code path seems to be untested.

chenghao-guo replied:
Sure

wjones127 reviewed (approved):
Seems alright. I have some minor optional feedback.

yanghua commented:
Since there are two approvers. So I will merge it.
